src/modules/Game.class.js

```diff
@@ -1,68 +1,265 @@
-
-  /**
-   * @returns {number[][]}
-   */
-  getState() {}
-
-  /**
-   * Returns the current game status.
-   *
-   * @returns {string} One of: 'idle', 'playing', 'win', 'lose'
-   *
-   * `idle` - the game has not started yet (the initial state);
-   * `playing` - the game is in progress;
-   * `win` - the game is won;
-   * `lose` - the game is lost
-   */
-  getStatus() {}
-
-  /**
-   * Starts the game.
-   */
-  start() {}
-
-  /**
-   * Resets the game.
-   */
-  restart() {}
-
-  // Add your own methods here
+  constructor(initialState = null) {
+    this.board = initialState || this.createEmptyBoard();
+    this.score = 0;
+    this.status = 'running';
+    this.mergedThisMove = [];
+  }
+
+  // Create an empty 4x4 board
+  createEmptyBoard() {
+    return Array.from({ length: 4 }, () => Array(4).fill(0));
+  }
+
+  // Get the current game state (board and score)
+  getState() {
+    return {
+      board: this.board,
+      score: this.score,
+      status: this.status,
+    };
+  }
+
+  // Get the current score
+  getScore() {
+    return this.score;
+  }
+
+  // Get the current status (running, won, game over)
+  getStatus() {
+    return this.status;
+  }
+
+  // Start the game (initialize the board and reset score)
+  start() {
+    this.board = this.createEmptyBoard();
+    this.score = 0;
+    this.status = 'running';
+    this.mergedThisMove = [];
+    this.addRandomTile();
+    this.addRandomTile();
+    this.updateUI();
+  }
+
+  // Restart the game
+  restart() {
+    this.start();
+  }
+
+  // Add a random tile (2 or 4) in a random empty cell
+  addRandomTile() {
+    const emptyCells = [];
+
+    // Loop to find all empty cells
+    for (let i = 0; i < 4; i++) {
+      for (let j = 0; j < 4; j++) {
+        if (this.board[i][j] === 0) {
+          emptyCells.push([i, j]);
+        }
+      }
+    }
+
+    // If there are any empty cells, place a random tile
+    if (emptyCells.length > 0) {
+      const [x, y] = emptyCells[Math.floor(Math.random() * emptyCells.length)];
+      const newValue = Math.random() < 0.9 ? 2 : 4;
+
+      this.board[x][y] = newValue;
+    }
+  }
+
+  // Check if a move is valid (if any tile moves or merges)
+  isValidMove() {
+    return this.board.some((row) => row.some((cell) => cell === 0));
+  }
+
+  // Check win or game over
+  checkGameStatus() {
+    if (this.board.some((row) => row.includes(2048))) {
+      this.status = 'won';
+      this.updateUI();
+      alert('You win!');
+    } else if (!this.isValidMove()) {
+      this.status = 'game over';
+      this.updateUI();
+      alert('Game Over!');
+    }
+  }
+
+  // Move left
+  moveLeft() {
+    let moved = false;
+
+    this.board.forEach((row, rowIndex) => {
+      const newRow = this.slideAndMerge(row);
+
+      if (newRow !== row) {
+        moved = true;
+      }
+      this.board[rowIndex] = newRow;
+    });
+
+    if (moved) {
+      this.addRandomTile();
+    }
+
+    this.checkGameStatus();
+    this.updateUI();
```

Comment:
This logic is similar in every method, try to put it in a helper function and use it everywhere.